src/OrchardCore.Modules/OrchardCore.Users/Services/DefaultControllerService.cs

```diff
@@ -0,0 +1,133 @@
+            ILogger<DefaultControllerService> logger)
+        {
+            _smtpService = smtpService;
+            _displayHelper = displayHelper;
+            _htmlEncoder = htmlEncoder;
+            _registrationFormEvents = registrationFormEvents;
+            _userService = userService;
+            _siteService = siteService;
+            _signInManager = signInManager;
+            _userManager = userManager;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Returns the created user, otherwise returns null
+        /// </summary>
+        /// <param name="controller"></param>
+        /// <param name="model"></param>
+        /// <param name="confirmationEmailSubject"></param>
+        /// <returns></returns>
+        public virtual async Task<IUser> RegisterUser(Controller controller, RegisterViewModel model, string confirmationEmailSubject)
+        {
+            var settings = (await _siteService.GetSiteSettingsAsync()).As<RegistrationSettings>();
+
+            if (settings.UsersCanRegister != UserRegistrationType.NoRegistration)
+            {
+                await _registrationFormEvents.InvokeAsync((e, modelState) => e.RegistrationValidationAsync((key, message) => modelState.AddModelError(key, message)), controller.ModelState, _logger);
+
+                if (controller.ModelState.IsValid)
+                {
+                    var user = await _userService.CreateUserAsync(new User { UserName = model.UserName, Email = model.Email, EmailConfirmed = !settings.UsersMustValidateEmail, IsEnabled = !settings.UsersAreModerated }, model.Password, (key, message) => controller.ModelState.AddModelError(key, message)) as User;
+
+                    if (user != null && controller.ModelState.IsValid)
+                    {
+                        if (settings.UsersMustValidateEmail && !user.EmailConfirmed)
+                        {
+                            // For more information on how to enable account confirmation and password reset please visit http://go.microsoft.com/fwlink/?LinkID=532713
+                            // Send an email with this link
+                            await SendEmailConfirmationTokenAsync(controller, user, confirmationEmailSubject);
+                        }
+                        else if (!(settings.UsersAreModerated && !user.IsEnabled))
+                        {
+                            await _signInManager.SignInAsync(user, isPersistent: false);
+                        }
+
+                        _logger.LogInformation(3, "User created a new account with password.");
+
+                        await _registrationFormEvents.InvokeAsync((e, user) => e.RegisteredAsync(user), user, _logger);
+
+                        return user;
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        public virtual async Task<bool> SendEmailAsync(string email, string subject, IShape model)
```

Comment:
Maybe move to an email service because it knows how to send an email based on a shape. Can be useful somewhere else.

Reply:
I though about that before, so let me organize this PR